File: other-languages/sorting/mergesort/mergesort.c
/*
Mergesort
Autor:
	John von Neumann, em 1945
Colaborador:
	Luiz Fernando Oliveira Corte Real (luiz@vidageek.net)
Tipo:
	sorting
Descrição:
	O algoritmo ordena um vetor dividindo-o pela metade e, depois de processar
	cada metade recursivamente, intercala as metades ordenadas.
Complexidade:
	O(n lgn)
Dificuldade:
	fácil
Referências:
	[1] http://en.wikipedia.org/wiki/Merge_sort
*/

#include <stdio.h>
#define TAMANHO_MAXIMO 100000

/* Intercala duas metades ordenadas do vetor v: de inicio até meio-1 e de meio até fim-1 */
void intercala(int v[], int inicio, int meio, int fim) {
	int vetor_auxiliar[TAMANHO_MAXIMO];
	int indice_parte_1, indice_parte_2, indice_vetor_auxiliar;
	indice_parte_1 = inicio;
	indice_parte_2 = meio;
	indice_vetor_auxiliar = 0;
	/* Mescla as metades até uma delas acabar */
	while (indice_parte_1 < meio && indice_parte_2 < fim) {
		if (v[indice_parte_1] < v[indice_parte_2]) {
			vetor_auxiliar[indice_vetor_auxiliar] = v[indice_parte_1];
			indice_parte_1++;
		} else {
			vetor_auxiliar[indice_vetor_auxiliar] = v[indice_parte_2];
			indice_parte_2++;
		}
		indice_vetor_auxiliar++;
	}
	/* Uma das metades pode ter ficado com elementos sobrando; apenas copiamos esses elementos */
	while (indice_parte_1 < meio) {
		vetor_auxiliar[indice_vetor_auxiliar] = v[indice_parte_1];
		indice_parte_1++;
		indice_vetor_auxiliar++;
	}
	while (indice_parte_2 < meio) {
		vetor_auxiliar[indice_vetor_auxiliar] = v[indice_parte_2];
		indice_parte_2++;
		indice_vetor_auxiliar++;
	}
	/* Copia o vetor auxiliar no lugar certo em v (de trás pra frente para não
	 * ter que calcular o tamanho do pedaço que vamos copiar; esse tamanho já
	 * está na variável indice_vetor_auxiliar) */
	while (indice_vetor_auxiliar > 0) {
		indice_vetor_auxiliar--;
		v[inicio + indice_vetor_auxiliar] = vetor_auxiliar[indice_vetor_auxiliar];
	}
}

/* Ordena o vetor v de inicio até fim-1 */
void mergesort_recursivo(int v[], int inicio, int fim) {
	int meio = (inicio + fim) / 2;
	/* Só chama a recursão na primeira metade se houver mais de um elemento lá */
	if (inicio < meio - 1) {
		mergesort_recursivo(v, inicio, meio);
	}
	/* Idem para a segunda metade */
	if (meio < fim - 1) {
		mergesort_recursivo(v, meio, fim);
	}
	intercala(v, inicio, meio, fim);
}

void mergesort(int v[], int n) {
	mergesort_recursivo(v, 0, n);
}

int main() {
	int vetor[] = { 5, 14, 8, 17, 1, -3, 9};
	int n = 7; /* número de elementos no vetor */
	int i;
	mergesort(vetor, n);
	printf("vetor ordenado:");
	for (i = 0; i < n; i++) {
		printf(" %d", vetor[i]);
	}
	printf("\n");
	return 0;
}

File: other-languages/sorting/mergesort/mergesort.rb

=begin
  Mergesort
Autor:
    ?
Colaborador:
    José Alberto O. Morais Filho (j.moraisg12@gmail.com)
Tipo:
    sorting
Descrição:
    Um dos melhores algoritmos de ordenação existente. Recebe um vetor, e, recursivamente, aplica a função Merge nos
  subvetores. A função merge recebe dois subvetores e os agrupa em um array ordenado.
Complexidade:
    O(n log n)
Dificuldade:
    médio
Referências:
   [1] Cormem, Thomas H. Introduction to Algorithms, 3rd Edition. (Páginas 29-34)
=end

def merge(array, left, mid, right)
  n1 = mid - left + 1
  n2 = right - mid

  l = Array.new(n1 + 1) {0}
  r = Array.new(n2 + 1) {0}

  for i in 0...n1 do
    l[i] = array[left + i]
  end

  for j in 1..n2 do
    r[j - 1] = array[mid + j]
  end

  l[n1] = 999999
  r[n2] = 999999

  i = 0
  j = 0

  for k in left..right do
    if l[i] <= r[j] then
      array[k] = l[i]
      i += 1
    else
      array[k] = r[j]
      j += 1
    end
  end
end

def mergesort(array, left, right)
  if left < right then
    mid = (left + right) / 2
    mergesort(array, left, mid)
    mergesort(array, mid + 1, right)
    merge(array, left, mid, right)
  end
end

a = [5, 3, 4, 1, 9, 7, 2, 6, 8]
mergesort(a, 0, 8)
print a

puts

b = [8, 1, 3, 2, 5, 7, 6, 4]
mergesort(b, 0, 7)
